clients/drcachesim/reader/snappy_file_reader.cpp

```diff
@@ -0,0 +1,312 @@
+        uncompressed_buf[size] = '\0';
+        ERRMSG("Unknown file type, got magic %s, want %s\n", uncompressed_buf.data(),
+               kMagic);
+        return false;
+    }
+    seen_magic = true;
+    return true;
+}
+
+bool
+snappy_reader_t::check_magic()
+{
+    if (!seen_magic) {
+        ERRMSG("Unknown file type, must start with magic chunk %s\n", kMagic);
+        return false;
+    }
+    return true;
+}
+
+bool
+snappy_reader_t::read_data_chunk(uint32_t size, chunk_type_t type)
+{
+    char *buf =
+        (type == COMPRESSED_DATA) ? compressed_buf.data() : uncompressed_buf.data();
+    size_t max_size = (type == COMPRESSED_DATA) ? kMaxCompressedSize : kMaxBlockSize;
+    max_size += kChecksumSize;
+    if (size < kChecksumSize || size > max_size) {
+        ERRMSG("Corrupted chunk header. Size %u, want <= %lu >= %lu.\n", size, max_size,
+               kChecksumSize);
+        return false;
+    }
+
+    fstream->read(buf, size);
+    if (!(*fstream)) {
+        return false;
+    }
+
+    uint32_t read_checksum = 0;
+    memcpy(&read_checksum, buf, 4);
+
+    src.reset(new snappy::ByteArraySource(&uncompressed_buf[kChecksumSize],
+                                          size - kChecksumSize));
+
+    // Potentially uncompress.
+    if (type == COMPRESSED_DATA) {
+        size_t uncompressed_chunk_size;
+        if (!snappy::GetUncompressedLength(&compressed_buf[kChecksumSize],
+                                           size - kChecksumSize,
+                                           &uncompressed_chunk_size)) {
+            ERRMSG("Failed getting snappy-compressed chunk length.\n");
+            return false;
+        }
+        if (uncompressed_chunk_size > kMaxBlockSize) {
+            ERRMSG("Uncompressed chunk larger than maximum size. Want <= %lu, got %lu\n",
+                   kMaxBlockSize, uncompressed_chunk_size);
+            return false;
+        }
+
+        if (!snappy::RawUncompress(&compressed_buf[kChecksumSize], size - kChecksumSize,
+                                   uncompressed_buf.data())) {
+            ERRMSG("Failed decompressing snappy chunk\n");
+            return false;
+        }
+        src.reset(new snappy::ByteArraySource(uncompressed_buf.data(),
+                                              uncompressed_chunk_size));
+    }
+
+    size_t dummy;
+    uint32_t checksum = mask_crc32(crc32c(src->Peek(&dummy), src->Available()));
+    if (checksum != read_checksum) {
+        ERRMSG("Checksum failure on snappy block. Want %x, got %x\n", read_checksum,
+               checksum);
+        return false;
+    }
+    return true;
+}
+
+bool
+snappy_reader_t::read_new_chunk()
+{
+    char buf[4];
+    fstream->read(buf, 4);
+    if (!(*fstream)) {
+        return false;
+    }
+    chunk_type_t type = static_cast<chunk_type_t>(buf[0]);
+    uint32_t size = 0;
+    memcpy(&size, buf + 1, 3);
+
+    switch (type) {
+    case STREAM_IDENTIFIER: {
+        return read_magic(size);
+    } break;
```

Comment:
This just looks unusual to me: I didn't know clang-format would sneak the `break` in on the same line as the `}`.  Given that we have no other code that looks like this I'm suggesting changing it.  I would vote for removing the braces and the break: simplify.  Ditto below.

Reply:
Done.